add public/internal/spi annotations

andimarek opened the pull request:
This is a first idea how we could handle #284:

There are three different Annotations: 
-`@PublicApi` for public classes/methods/interfaces.
- `@PrivateApi` for private stuff
- `@PublicSpi` for public interfaces/abstract classes which users are expected to implement.

bbakerman commented:
Naming is hard.

I would use @Internal rather than @InternalApi

Also I think we should consider @ExperimentalApi and ExperimentalSpi.  This allows you to stage new code without the super strong commitment.

bbakerman commented on src/main/java/graphql/PublicApi.java:
```diff
@@ -0,0 +1,14 @@
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.ElementType.TYPE;
+
+@Retention(RetentionPolicy.RUNTIME)
```
Consider adding

java.lang.annotation.Documented

@Documented

bbakerman commented on src/main/java/graphql/PublicSpi.java:
```diff
@@ -0,0 +1,14 @@
+import static java.lang.annotation.ElementType.CONSTRUCTOR;
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.ElementType.TYPE;
+
```
Consider @Documented

bbakerman commented on src/main/java/graphql/GraphQL.java:
```diff
@@ -89,6 +93,7 @@ public GraphQL(GraphQLSchema graphQLSchema, ExecutionStrategy queryStrategy, Exe
+    @InternalApi
     public GraphQL(GraphQLSchema graphQLSchema, ExecutionStrategy queryStrategy, ExecutionStrategy mutationStrategy, ExecutionStrategy subscriptionStrategy) {
         this(graphQLSchema, queryStrategy, mutationStrategy, subscriptionStrategy, DEFAULT_EXECUTION_ID_PROVIDER, NoOpInstrumentation.INSTANCE);
     }
```
In 4.0 we should remove deprecations.

andimarek commented:
@bbakerman Changed it to `@Internal` add added `@Documented`.

I am not really sure about the "rules" where to put these annotations: I added the on the class/interface, but not on every method. I quietly assume that every method is public or internal if the type itself is so. If this is not the case I added a extra annotation: for example the constructor is marked as `internal`, but the class itself is `public`.

bbakerman commented:
Yeah I think you put them on the class and then IF you want to make an exemption, you put that into the method.

The typical ones will be constructors.  By making them @internal you allow them to be expanded over time without breaking API.  The builders are how people should construct them and hence we can default values there etc..

andimarek commented:
It is a first start: we will add new annotations before 3.0.0
